src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java

```diff
@@ -186,21 +186,16 @@ public StorageCalculator(boolean forArguments) {
             this.forArguments = forArguments;
         }
 
-        VMStorage stackAlloc(long size, long alignment) {
+        void alignStack(long alignment) {
+            stackOffset = Utils.alignUp(stackOffset, alignment);
+        }
+
+        VMStorage stackAlloc(long size, long alignment, boolean subSlotPacking) {
             assert forArguments : "no stack returns";
-            // Implementation limit: each arg must take up at least an 8 byte stack slot (on the Java side)
-            // There is currently no way to address stack offsets that are not multiples of 8 bytes
-            // The VM can only address multiple-of-4-bytes offsets, which is also not good enough for some ABIs
-            // see JDK-8283462 and related issues
-            long stackSlotAlignment = Math.max(alignment, STACK_SLOT_SIZE);
+            long stackSlotAlignment = subSlotPacking && !forVarArgs
+                    ? alignment
+                    : Math.max(alignment, STACK_SLOT_SIZE);
```

Comment:
With spillStruct(Un)box, and `nextStorageForHFA` now also accounting for sub slot packing, I think that maybe this is not the right level to adjust the alignment (again).

I suggest removing the conditional expression here, and simply using the `alignment` passed to this method in place of `stackSlotAlignment`. The alignment can be adjusted based on `requiresSubSlotStackPacking()` in the other  `stackAlloc(MemorryLayout)` overload instead, which looks to be the only place that doesn't pass a constant boolean value. Then, `nextStorageForHFA` can also just call that overload for each member layout.